File: cli/src/providers/anthropic.js
// Anthropic Messages API — streaming (SSE) with tool use.
// The API key goes directly to api.anthropic.com (or cfg.anthropicBaseUrl).
import { createSSE } from "../util.js";

export class ProviderError extends Error {}

/**
 * cfg: resolved config
 * params: { system, messages (internal history), tools, signal }
 * events: { onText(delta), onToolStart(name, id) }
 * returns: { text, toolCalls: [{id, name, input}], stopReason, usage }
 */
export async function chatStream(cfg, params, events = {}) {
  const key = cfg.anthropicKey;
  if (!key) throw new ProviderError("No Anthropic API key. Run `roforge login` or set ANTHROPIC_API_KEY.");
  const body = {
    model: params.model,
    max_tokens: params.maxTokens || cfg.maxTokens || 8000,
    stream: true,
    messages: renderHistory(params.messages),
  };
  if (params.system) body.system = params.system;
  if (params.tools && params.tools.length) body.tools = renderTools(params.tools);

  let res;
  try {
    res = await fetch(`${cfg.anthropicBaseUrl}/v1/messages`, {
      method: "POST",
      headers: {
        "x-api-key": key,
        "anthropic-version": "2023-06-01",
        "content-type": "application/json",
        accept: "text/event-stream",
      },
      body: JSON.stringify(body),
      signal: params.signal,
    });
  } catch (e) {
    throw new ProviderError(`network error calling Anthropic: ${e.cause?.code || e.message}`);
  }

  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new ProviderError(`Anthropic HTTP ${res.status}: ${text.slice(0, 400)}`);
  }

  let text = "";
  let stopReason = null;
  let usage = { input_tokens: 0, output_tokens: 0 };
  const toolCalls = [];
  let currentTool = null;

  const sse = createSSE((ev) => {
    const d = ev.data;
    if (typeof d !== "object" || !d) return;
    switch (ev.event) {
      case "message_start":
        if (d.message && d.message.usage) usage = { ...usage, ...d.message.usage };
        break;
      case "content_block_start":
        if (d.content_block && d.content_block.type === "tool_use") {
          currentTool = { id: d.content_block.id, name: d.content_block.name, input: {}, _json: "" };
          toolCalls.push(currentTool);
          events.onToolStart && events.onToolStart(currentTool.name, currentTool.id);
        }
        break;
      case "content_block_delta": {
        const delta = d.delta;
        if (!delta) break;
        if (delta.type === "text_delta") {
          text += delta.text || "";
          events.onText && events.onText(delta.text || "");
        } else if (delta.type === "input_json_delta" && currentTool) {
          currentTool._json += delta.partial_json || "";
        }
        break;
      }
      case "content_block_stop":
        if (currentTool) {
          try {
            currentTool.input = currentTool._json ? JSON.parse(currentTool._json) : {};
          } catch {
            currentTool.input = {};
          }
          delete currentTool._json;
          currentTool = null;
        }
        break;
      case "message_delta":
        if (d.delta && d.delta.stop_reason) stopReason = d.delta.stop_reason;
        if (d.usage) usage = { ...usage, ...d.usage };
        break;
      default:
        break;
    }
  });

  const reader = res.body.getReader();
  const decoder = new TextDecoder();
  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    sse.push(decoder.decode(value, { stream: true }));
  }
  sse.end();

  return { text, toolCalls, stopReason, usage };
}

// Internal history → Anthropic messages.
// items: {role:"user",text} | {role:"assistant",text?,calls?} | {role:"tool",id,name,result}
export function renderHistory(history) {
  const msgs = [];
  for (const item of history) {
    if (item.role === "user") {
      msgs.push({ role: "user", content: item.text });
    } else if (item.role === "assistant") {
      const blocks = [];
      if (item.text) blocks.push({ type: "text", text: item.text });
      if (item.calls) {
        for (const c of item.calls) {
          blocks.push({ type: "tool_use", id: c.id, name: c.name, input: c.args || c.input || {} });
        }
      }
      if (blocks.length) msgs.push({ role: "assistant", content: blocks });
    } else if (item.role === "tool") {
      // content: string, or [image block, text block] when the tool returned
      // a vision capture (e.g. forge_viewport).
      let content = item.result;
      if (item.image && item.image.base64) {
        content = [
          {
            type: "image",
            source: {
              type: "base64",
              media_type: item.image.mediaType || "image/png",
              data: item.image.base64,
            },
          },
          { type: "text", text: item.result },
        ];
      }
      const block = { type: "tool_result", tool_use_id: item.id, content };
      const last = msgs[msgs.length - 1];
      if (last && last.role === "user" && Array.isArray(last.content)) {
        last.content.push(block);
      } else {
        msgs.push({ role: "user", content: [block] });
      }
    }
  }
  return msgs;
}

export function renderTools(tools) {
  return tools.map((t) => ({
    name: t.name,
    description: t.description,
    input_schema: t.inputSchema || t.input_schema,
  }));
}

File: client/src/Root/RoForge/Providers/Anthropic.lua
-- Anthropic Messages API adapter.
-- Direct call from Studio: the x-api-key header goes straight to
-- api.anthropic.com. It never touches the RoForge backend or any proxy.

local Http = require(script.Parent.Parent.Http)

local Anthropic = {}

Anthropic.name = "anthropic"
Anthropic.defaultModel = "claude-sonnet-4-5" -- edit in Settings to use a newer model id

-- Internal history -> Anthropic messages (tool results merge into user turns).
local function renderHistory(history)
	local msgs = {}
	for _, item in ipairs(history) do
		if item.role == "user" then
			table.insert(msgs, { role = "user", content = item.text })
		elseif item.role == "assistant" then
			local blocks = {}
			if item.text ~= nil and item.text ~= "" then
				table.insert(blocks, { type = "text", text = item.text })
			end
			if item.calls then
				for _, call in ipairs(item.calls) do
					table.insert(blocks, {
						type = "tool_use",
						id = call.id,
						name = call.name,
						input = call.args or {},
					})
				end
			end
			if #blocks > 0 then
				table.insert(msgs, { role = "assistant", content = blocks })
			end
		else -- role == "tool"
			local block = {
				type = "tool_result",
				tool_use_id = item.id,
				content = item.result,
			}
			local last = msgs[#msgs]
			if last and last.role == "user" and type(last.content) == "table" then
				table.insert(last.content, block)
			else
				table.insert(msgs, { role = "user", content = { block } })
			end
		end
	end
	return msgs
end

local function renderTools(tools)
	local out = {}
	for _, t in ipairs(tools) do
		table.insert(out, {
			name = t.name,
			description = t.description,
			input_schema = t.input_schema,
		})
	end
	return out
end

function Anthropic.chat(cfg, system, history, tools)
	local model = (cfg.Model and cfg.Model ~= "") and cfg.Model or Anthropic.defaultModel
	local headers = {
		["x-api-key"] = cfg.ApiKey,
		["anthropic-version"] = "2023-06-01",
		["content-type"] = "application/json",
	}
	local body = {
		model = model,
		max_tokens = 4096,
		messages = renderHistory(history),
	}
	if system and system ~= "" then
		body.system = system
	end
	if tools and #tools > 0 then
		body.tools = renderTools(tools)
	end

	local resp = Http.postJson("https://api.anthropic.com/v1/messages", headers, body)
	if not resp.Success or (resp.StatusCode and resp.StatusCode >= 400) then
		return nil, Http.errorText(resp)
	end
	local data, err = Http.decode(resp.Body)
	if err then
		return nil, err
	end

	local text = ""
	local calls = {}
	for _, block in ipairs(data.content or {}) do
		if block.type == "text" then
			text = text .. tostring(block.text or "")
		elseif block.type == "tool_use" then
			table.insert(calls, { id = block.id, name = block.name, args = block.input or {} })
		end
	end
	return { text = text, calls = calls }, nil
end

return Anthropic
